src/lowlevel_api.jl

```diff
@@ -0,0 +1,258 @@
+    elseif type_==:Bridge
+        error("not implemented")
+    elseif type_==:PCI_Device
+        error("not implemented")
+    elseif type_==:OS_Device
+        error("not implemented")
+    elseif type_==:Die
+        ha = unsafe_load(convert(Ptr{hwloc_cache_attr_s}, hattr))
+        return DieAttr(ha.depth)
+    elseif type_==:MemCache
+        error("not implemented")
+    else
+        error("Unsupported object type $type_")
+    end
+end
+
+
+
+
+struct Object
+    type_::Symbol
+    os_index::Int
+    name::String
+    attr::Attribute
+    mem::Int64
+
+    depth::Int
+    logical_index::Int
+    # os_level::Int
+
+    children::Vector{Object}
+
+    memory_children::Vector{Object}
+
+    # Object() = new(:Error, -1, "(nothing)", NullAttr(),
+    #                0, -1, -1, # -1,
+    #                Object[], Object[])
+end
+
+show(io::IO, obj::Object) = print(io, "Hwloc.Object: $(obj.type_)")
+
+IteratorSize(::Type{Object}) = Base.SizeUnknown()
+IteratorEltype(::Type{Object}) = Base.HasEltype()
+eltype(::Type{Object}) = Object
+isempty(::Object) = false
+iterate(obj::Object) = (obj, isempty(obj.memory_children) ? obj.children : vcat(obj.memory_children, obj.children))
+function iterate(::Object, state::Vector{Object})
+    isempty(state) && return nothing
+    # depth-first traversal
+    # obj = shift!(state)
+    obj, state = state[1], state[2:end]
+    prepend!(state, obj.children)
+    prepend!(state, obj.memory_children)
+    return obj, state
+end
+# length(obj::Object) = mapreduce(x->1, +, obj)
+
+attributes(obj::Object) = obj.attr
+children(obj::Object) = obj.children
+
+
+# Load topology for an object and all its children
+function load(hobj::hwloc_obj_t)
+    @assert hobj != C_NULL
+    obj = unsafe_load(hobj)
+
+    @assert Integer(obj.type)>=0 && Integer(obj.type)<length(obj_types)
+    type_ = obj_types[obj.type+1]
+
+    os_index = mod(obj.os_index, Cint)
+
+    name = obj.name == C_NULL ? "" : unsafe_string(obj.name)
+
+    attr = load_attr(obj.attr, type_)
+
+    mem = Int64(obj.total_memory)
+
+    depth = obj.depth
+
+    logical_index = obj.logical_index
+
+    # topo.os_level = obj.os_level
+
+    obj_children = Vector{hwloc_obj_t}(UndefInitializer(), obj.arity)
+    ccall(:memcpy, Ptr{Cvoid}, (Ptr{Cvoid}, Ptr{Cvoid}, Csize_t), obj_children,
+          obj.children, obj.arity*sizeof(Ptr{Cvoid}))
+
+    children = Object[load(child) for child in obj_children]
+
+    memory_children = Object[]
+    if obj.memory_arity != C_NULL && obj.memory_first_child != C_NULL
+        push!(memory_children, load(obj.memory_first_child))
+    end
+
+    topo = Object(type_, os_index, name, attr, mem, depth, logical_index, children, memory_children)
+    return topo
+end
+
+
+"""
+    topology_load() -> Hwloc.Object
+
+Load the system topology by calling into libhwloc.
+"""
+function topology_load()
+    htopop = Ref{Ptr{Cvoid}}()
+    ierr = ccall((:hwloc_topology_init, libhwloc), Cint, (Ptr{Cvoid},), htopop)
```

Comment:
Should this no use the new wrappers?

Reply:
Crap! I was distracted -- lemme fix that now

Reply:
Done! -- Waiting for CI.

Reply:
Since we're on the topic of ccall -- what's the "Julianic" way to call `memcpy`? `ccall(:memcpy ...` or are there wrappers for this somewhere already?

Reply:
Why do you want to call `memcpy` directly and not use, e.g., `copyto!`?

FWIW, the latter calls [`jl_array_copy`](https://github.com/JuliaLang/julia/blob/46135dfce9074e5bf94eb277de28a33cad9cc14f/src/array.c#L1176-L1189) which, among other things, calls `memcpy`.

Reply:
Done: https://github.com/JuliaParallel/Hwloc.jl/pull/64/commits/b2928b0bb0682bda2688c9c14f19cbe6122c83fb

No more ccall

Reply:
And of course I messed up and it segfaults in CI (not my laptop though :thinking: ) -- Anyway, fixed it.